sw/device/tests/alert_handler_lpg_sleep_mode_alerts.c

```diff
@@ -0,0 +1,528 @@
+    num_fired_loc_alerts = alert_handler_num_fired_loc_alerts();
+    LOG_INFO("num_fired alerts = %d, num_fired_loc_alerts = %d",
+             num_fired_alerts, num_fired_loc_alerts);
+
+    // Check if the number of the fired alerts/local-alerts are correct
+    CHECK(num_fired_alerts == 1,
+          "Phase #1: Only a single alert must be fired after wakeup!");
+    CHECK(num_fired_loc_alerts == 0,
+          "Phase #1: No local alerts must be fired!");
+
+    // Check if the expected fatal alert has been fired.
+    CHECK_DIF_OK(dif_alert_handler_alert_is_cause(
+        &alert_handler, kExpectedAlertNumber, &is_cause));
+    CHECK(is_cause, "Phase #1: Expected alert has NOT been fired!!");
+
+    // Increment the test_step counter for the next test step
+    flash_ctrl_testutils_counter_increment(&flash_ctrl, kCounterTestSteps);
+    test_step_cnt = flash_ctrl_testutils_counter_get(kCounterTestSteps);
+
+    // Set the AON timer to send a wakeup signal in ~10-50us.
+    aon_timer_testutils_wakeup_config(&aon_timer,
+                                      rand_testutils_gen32_range(2, 10));
+    // Enter the normal sleep mode
+    enter_low_power(/*deep_sleep=*/false);
+  } else {
+    dif_pwrmgr_wakeup_reason_t wakeup_reason;
+    CHECK_DIF_OK(dif_pwrmgr_wakeup_reason_get(&pwrmgr, &wakeup_reason));
+    LOG_ERROR("Unexpected wakeup_reason=%x", wakeup_reason);
+    CHECK(false, "Unexpected wakeup reason");
+  }
+}
+
+/**
+ * External interrupt handler.
+ */
+void ottf_external_isr(void) {
+  dif_pwrmgr_irq_t irq_id;
+  top_earlgrey_plic_peripheral_t peripheral;
+
+  isr_testutils_pwrmgr_isr(plic_ctx, pwrmgr_isr_ctx, &peripheral, &irq_id);
+
+  // Check that both the peripheral and the irq id is correct
+  CHECK(peripheral == kTopEarlgreyPlicPeripheralPwrmgrAon,
+        "IRQ peripheral: %d is incorrect", peripheral);
+  CHECK(irq_id == kDifPwrmgrIrqWakeup, "IRQ ID: %d is incorrect", irq_id);
+}
+
+bool test_main(void) {
+  bool is_cause;
+  bool is_locked;
+  bool sleep_mode;
+  uint16_t num_fired_alerts;
+  uint16_t num_fired_loc_alerts;
+
+  init_test_components();
+
+  // TEST PHASE #1 (ping-timeout = 256)
+  // To ensure that fatal alerts keep firing after sleep/wakeup.
+  // Only a single fatal alert is expected to be fired.
+  // No local alerts is expected to be fired.
+  while (test_step_cnt < num_total_wakeups) {
+    // TODO: It seems to be that the only way to continue the SW execution
+    // after a sram_ctrl or flash_ctrl fatal fault is a reset.
+    // In this tes, we are only interested in the shallow sleep mode.
+    // For checking sram_ctrl and flash_ctrl fatal alerts we may need an SV
+    // sequence.
+    if (kExpectedAlertNumber == kTopEarlgreyAlertIdFlashCtrlFatalStdErr ||
+        kExpectedAlertNumber == kTopEarlgreyAlertIdSramCtrlMainFatalError) {
+      LOG_INFO(
+          "No way to continue after Flash_ctrl or Sram_ctrl fatal errors!!");
+      break;
+    }
+
```

Comment:
Is there a good way to handle sram_ctrl and flash_ctrl fatal alerts for this test?

Reply:
I don't think so, no. If flash_ctrl escalates, instruction fetches will get stuck. If the main SRAM escalates, any accesses to variables on the heap or stack will error out.

Reply:
@msfschaffner, Thanks a lot Michael! 

Cindy will take a look at it to see how we can handle these alerts

Reply:
I think one solution here is to use a plusarg to not include `sram` and `flash_ctrl` fault injections.
I was thinking something like this in the hjson:
```
     {
      name: chip_sw_alert_handler_lpg_sleep_mode_alerts
      uvm_test_seq: chip_sw_all_escalation_resets_vseq
      sw_images: ["//sw/device/tests:alert_handler_lpg_sleep_mode_alerts_test:1"] 
      en_run_modes: ["sw_test_mode_test_rom"]
      run_opts: ["+en_scb=0", "+sw_test_timeout_ns=3000_000_000", "+bypass_alert_ready_to_end_check=1", "+fatal_inject_exclude=sram_ctrl,flash_ctrl"]
      run_timeout_mins: 240
      reseed: 60
      }
```

Then in `chip_sw_all_escalation_resets_vseq`, take the plusarg and find the IP names to exclude.

If you think this is an okay solution, I can help with the DV part.

Reply:
Honestly, handling this in C would be ok for now as well. We do the same thing in another, related all_escalations_resets test.

Reply:
@msf, I agree the ideal solution would be to handle it in SW but I couldn't come with a solution so far. In `all_escalation_resets` test, they configure the alert handler such that it will generate escalation alerts. After the escalation reset, they can check if the cause was flash_ctrl/sram_ctrl error: they keep counter to count number of serviced interrupts in flash. If sram_ctrl/flash_ctrl was received, interrupts does not get serviced.

@msfschaffner, @cindy, For M2, would it be okay to go with the SV method that Cindy suggested if we couldn't find a easy software fix?

Reply:
Yeah that is totally fine for me.

Reply:
Sorry, what I meant was just to keep the workaround if statement
```
    if (kExpectedAlertNumber == kTopEarlgreyAlertIdFlashCtrlFatalStdErr ||
        kExpectedAlertNumber == kTopEarlgreyAlertIdSramCtrlMainFatalError) {
      LOG_INFO(
          "No way to continue after Flash_ctrl or Sram_ctrl fatal errors!!");
      break;
    }
```
 in C for now so you can make progress - but if it is easy to add what @cindychip mentioned, we can do that as well.

I guess if you keep the workaround if statement above, some tests would just "fall through" and not test anything, right?

Reply:
@msfschaffner, Oh I see :) and yes, you are right about the workaround if statement; my intention was to make some tests just fall through until we come up with a nicer solution. 

With the current C implementation, I am able to end the C part but the simulation still gets stuck. I am trying to figure out if I can  trigger a deep-sleep/wakeup cycle such that the fatal alert will be cleared after the wakeup. 

And @cindychip is working on an SV solution.

Reply:
lmk if https://github.com/lowRISC/opentitan/pull/16442 works :)

Reply:
@cindychip I tested it on my DV environment (reseed:120) and it seems to be working, thanks!